ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py

```diff
@@ -240,6 +239,29 @@ def _clean_up(self):
             self.ssh_client.close()
         if self.instance_ids:
             self._connect().terminate_instances(InstanceIds=self.instance_ids)
+            waiter = self._connect().get_waiter('instance_terminated')
+            repeat_count = 1
+            error_msg = 'Instance did not stop within allotted time'
+            while repeat_count <= self.wait_count:
+                try:
+                    wait_status = waiter.wait(
+                        InstanceIds=[self.helper_instance['InstanceId']],
+                        Filters=[
+                            {
+                                'Name': 'instance-state-name',
+                                'Values': ['terminated']
+                            }
+                        ]
+                    )
+                except:
+                    wait_status = 1
+                if self.verbose:
+                    self.progress_timer.cancel()
+                repeat_count = self._check_wait_status(
+                    wait_status,
+                    error_msg,
+                    repeat_count
+                )
```

Comment:
This is correct, just not certain if we really need/want to wait for the instance to terminate. Volume detach and delete works even if the platform considers the instance running. The only thing this would do is "protect" the user from incurring costs if an instance fails to properly shut down. But I have not ever seen this. Well I have seen instances take a longtime to terminate, but they eventually all do. Thus I am not certain we need to wait here.

Below the check for "temporary_key_created should be removed as this is now in the EC2Setup class implementation.

Reply:
This is necessary because of deleting the temporary VPC and Security Group. If you want to delete the VPC / Security Group of a still running (or stopped) instance, you will get a DependentResource error. However, we could only do that in case there is a temporary resource but then either

a) we need to hand over to EC2ImageUploader if there is a temporary resource (don't like that)
or
b) do it in the EC2Setup class. But then we need to hand over the helper instance(s) ID(s) to the EC2Setup class (don't like that either)
So in either way, we need to handover information from one class to another. So I thought simplest solution is to always wait till the instance is properly terminated.

Reply:
agreed

Reply:
agreed